internal/cmd/commands/dev/dev.go

```diff
@@ -824,73 +823,84 @@ func (c *Command) Run(args []string) int {
 	c.opsServer = opsServer
 	c.opsServer.Start()
 
-	// Wait for shutdown
-	shutdownTriggered := false
-
-	// Add a force-shutdown goroutine to consume more interrupts
-	abortForceShutdownCh := make(chan struct{})
-	defer close(abortForceShutdownCh)
+	// On first interrupt, enter graceful shutdown for controllers and workers (workers wait for connections to drain)
+	// On second interrupt, force controller shutdown and send workers into shutdown (stop all connections)
+	// On third+ interrupt, force worker shutdown
+	shutdownDone := false
+	shutdownTriggerCount := uint32(0)
 
 	runShutdownLogic := func() {
-		go func() {
-			shutdownCh := make(chan os.Signal, 4)
-			signal.Notify(shutdownCh, os.Interrupt, syscall.SIGTERM)
-			for {
-				select {
-				case <-shutdownCh:
-					c.UI.Error("Forcing shutdown")
-					os.Exit(base.CommandUserError)
+		if sync.LoadUint32(&shutdownTriggerCount) == 1 {
+			go func() {
```

Comment:
The logic in here currently makes an assumption that `c.controller.Shutdown()` will only be called during the first SIGINT and that we will actually reach it -- but if we're stuck waiting on graceful worker shutdown we may not. We may want the second SIGINT to also trigger an immediate controller shutdown, and save the `os.Exit` for the third in all cases.

There would be some knock-on effects. Probably there would need to be separate goroutines for worker shutdown and controller shutdown to run concurrently to ensure we're not blocking controller shutdown at that point on worker shutdown -- something we _don't_ want in the graceful case! And that also means that `shutdownDone` becomes harder to understand -- right now it's set when the functions return, but what if we end up running controller shutdown concurrently and one ends up returning immediately with an error? Possibly the thing to do would be to only set `shutdownDone` if the shutdown func returns with no error.